languages/rust/src/polar.rs

```diff
@@ -0,0 +1,127 @@
+/// Communicate with the Polar virtual machine: load rules, make queries, etc/
+use std::fs::File;
+use std::io::Read;
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+
+use crate::host::Host;
+
+#[derive(Clone)]
+pub struct Polar {
+    inner: Rc<crate::PolarCore>,
+    host: Arc<Mutex<Host>>,
+}
+
+impl Default for Polar {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl Polar {
+    pub fn new() -> Self {
+        let inner = Rc::new(crate::PolarCore::new());
+        let mut host = Host::new(Rc::downgrade(&inner));
+
+        // register all builtin constants
+        for (name, class) in crate::host::builtins::constants() {
+            let value = host.cache_class(class, name.clone());
+            inner.register_constant(name, value);
+        }
+
+        Self {
+            host: Arc::new(Mutex::new(host)),
+            inner,
+        }
+    }
+
+    pub fn clear(&mut self) {
+        *self = Self::new();
+    }
+
+    fn check_messages(&mut self) {
+        while let Some(msg) = self.inner.next_message() {
+            match msg.kind {
+                polar_core::types::MessageKind::Print => tracing::trace!("{}", msg.msg),
+                polar_core::types::MessageKind::Warning => tracing::warn!("{}", msg.msg),
+            }
+        }
+    }
+
+    fn check_inline_queries(&mut self) -> anyhow::Result<()> {
```

Comment:
We should be using https://crates.io/crates/thiserror for library errors, and anyhow for the application (tests, repl + debugger).

We're using it here because it's super convenient and easy to get something initially working.